File: compare_strings.py
import sys
from typing import Tuple, Optional, TextIO

ALPHABET = {'A', 'C', 'G', 'N', 'T'}
MAX_SEQ_LEN = 1000

# ---------------------------------------------------------------------
# Helpers FASTA
# ---------------------------------------------------------------------

def fasta_open_and_read_header(path: str) -> Tuple[Optional[TextIO], Optional[str]]:
    """
    Ouvre un fichier FASTA et lit l'en-tête.
    Retourne (fp, name) où :
      - fp est un fichier ouvert en lecture, positionné juste après l'en-tête
      - name est le nom de séquence sans le '>'
    En cas d'erreur : affiche un message et retourne (None, None).
    """
    try:
        fp = open(path, "r")
    except OSError:
        print(f"Erreur: impossible d'ouvrir le fichier '{path}'.", file=sys.stderr)
        return None, None

    header = fp.readline()
    if header == "":
        print(f"Erreur: le fichier '{path}' est vide.", file=sys.stderr)
        fp.close()
        return None, None

    header = header.strip()
    if not header.startswith(">"):
        print(f"Erreur: format FASTA invalide dans '{path}' (en-tête manquant).", file=sys.stderr)
        fp.close()
        return None, None

    name = header[1:]
    if not name:
        print(f"Erreur: nom de séquence vide dans '{path}'.", file=sys.stderr)
        fp.close()
        return None, None

    return fp, name


def fasta_read_bases(fp: Optional[TextIO], n: int) -> Optional[str]:
    """
    Lit au maximum n bases depuis le fichier fp.
    Ignore les caractères '\\n' et '\\r'.

    Retourne une chaîne contenant au plus n caractères.
    Retourne "" si fin de séquence.
    Retourne None en cas d'erreur de format (ex: second '>').
    """
    if fp is None:
        return None

    bases = []

    while len(bases) < n:
        c = fp.read(1)

        if c == "":
            # EOF
            break

        if c == "\n" or c == "\r":
            continue

        if c == ">":
            print("Erreur: format FASTA invalide (plusieurs séquences détectées).", file=sys.stderr)
            return None

        bases.append(c)

    return "".join(bases)


# ---------------------------------------------------------------------
# Helpers CSV
# ---------------------------------------------------------------------

def csv_open(path: str) -> Optional[TextIO]:
    """
    Ouvre le fichier CSV de sortie en écriture.
    Retourne fp ou None si erreur.
    """
    try:
        return open(path, "w")
    except OSError:
        print(f"Erreur: impossible d'ouvrir le fichier de sortie '{path}'.", file=sys.stderr)
        return None


def csv_write_mutation(fp: TextIO, name: str, pos: int, ref: str, alt: str) -> None:
    """
    Écrit une substitution sur une ligne CSV.
    """
    fp.write(f"{name},{pos},{ref},{alt}\n")


def csv_close(fp: Optional[TextIO]) -> None:
    if fp is not None:
        fp.close()


# ---------------------------------------------------------------------
# Programme principal
# ---------------------------------------------------------------------

def main():
    if len(sys.argv) != 4:
        print("Usage: python compare_strings.py chemin_ref_seq chemin_seq chemin_output",
              file=sys.stderr)
        return 1

    path_ref = sys.argv[1]
    path_seq = sys.argv[2]
    path_out = sys.argv[3]

    # --- Ouverture des fichiers FASTA ---
    fp_ref, name_ref = fasta_open_and_read_header(path_ref)
    if fp_ref is None:
        return 1

    fp_seq, name_seq = fasta_open_and_read_header(path_seq)
    if fp_seq is None:
        fp_ref.close()
        return 1

    # Vérifier que les noms sont identiques
    if name_ref != name_seq:
        print(f"Erreur: les noms de séquences diffèrent ('{name_ref}' vs '{name_seq}').",
              file=sys.stderr)
        fp_ref.close()
        fp_seq.close()
        return 1

    # --- Ouverture du CSV ---
    fp_out = csv_open(path_out)
    if fp_out is None:
        fp_ref.close()
        fp_seq.close()
        return 1

    # --- Lecture des séquences (phase 1 : chargement complet autorisé) ---
    seq_ref = fasta_read_bases(fp_ref, MAX_SEQ_LEN)
    if seq_ref is None:
        csv_close(fp_out)
        fp_ref.close()
        fp_seq.close()
        return 1

    seq = fasta_read_bases(fp_seq, MAX_SEQ_LEN)
    if seq is None:
        csv_close(fp_out)
        fp_ref.close()
        fp_seq.close()
        return 1

    # Vérifier que les longueurs sont identiques
    if len(seq_ref) != len(seq):
        print(f"Erreur: les séquences ont des longueurs différentes ({len(seq_ref)} vs {len(seq)}).",
              file=sys.stderr)
        csv_close(fp_out)
        fp_ref.close()
        fp_seq.close()
        return 1

    # Vérifier l'alphabet et comparer les séquences
    for i in range(len(seq_ref)):
        if seq_ref[i] not in ALPHABET:
            print(f"Erreur: caractère invalide '{seq_ref[i]}' à la position {i} "
                  f"dans la séquence de référence.", file=sys.stderr)
            csv_close(fp_out)
            fp_ref.close()
            fp_seq.close()
            return 1
        if seq[i] not in ALPHABET:
            print(f"Erreur: caractère invalide '{seq[i]}' à la position {i} "
                  f"dans la séquence à comparer.", file=sys.stderr)
            csv_close(fp_out)
            fp_ref.close()
            fp_seq.close()
            return 1
        if seq_ref[i] != seq[i]:
            csv_write_mutation(fp_out, name_ref, i, seq_ref[i], seq[i])

    # --- Fermeture des fichiers ---
    csv_close(fp_out)
    fp_ref.close()
    fp_seq.close()

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: compare_strings.cpp
#include <cstdio>
#include <cstdlib>
#include <cstring>

#include "utils_fasta.h"
#include "utils_csv.h"


// Hypothèses de la phase 1 (voir énoncé)
static const size_t MAX_SEQ_NAME_LEN = 101; // 100 caractères + '\0'
static const size_t MAX_SEQ_LEN = 1000;     // max 1000 caractères

static const char ALPHABET[] = "ACGNT";

static bool is_valid_base(char c) {
    for (size_t i = 0; ALPHABET[i] != '\0'; ++i) {
        if (c == ALPHABET[i]) return true;
    }
    return false;
}

static void close_all(FILE* fp_ref, FILE* fp_seq, FILE* fp_out) {
    if (fp_ref) fasta_close(fp_ref);
    if (fp_seq) fasta_close(fp_seq);
    if (fp_out) csv_close(fp_out);
}


int main(int argc, char* argv[]) {
    
    if (argc != 4) {
        fprintf(stderr, "Usage: %s chemin_ref_seq chemin_seq chemin_output\n", argv[0]);
        return -1;
    }
    const char* path_ref = argv[1];
    const char* path_seq = argv[2];
    const char* path_out = argv[3];

    // --- Ouverture des fichiers FASTA ---
    FILE* fp_ref = nullptr;
    FILE* fp_seq = nullptr;
    FILE* fp_out = nullptr;

    char name_ref[MAX_SEQ_NAME_LEN];
    char name_seq[MAX_SEQ_NAME_LEN];

    FastaStatus fs;

    fs = fasta_open_and_read_header(path_ref, &fp_ref, name_ref, MAX_SEQ_NAME_LEN);
    if (fs != FASTA_OK) {
        fprintf(stderr, "Erreur: impossible d'ouvrir ou de lire l'en-tête du fichier de référence '%s' (code %d).\n", path_ref, fs);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    fs = fasta_open_and_read_header(path_seq, &fp_seq, name_seq, MAX_SEQ_NAME_LEN);
    if (fs != FASTA_OK) {
        fprintf(stderr, "Erreur: impossible d'ouvrir ou de lire l'en-tête du fichier de séquence '%s' (code %d).\n", path_seq, fs);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    // Vérifier que les noms sont identiques
    if (std::strcmp(name_ref, name_seq) != 0) {
        fprintf(stderr, "Erreur: les noms de séquences diffèrent ('%s' vs '%s').\n", name_ref, name_seq);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    // --- Ouverture du fichier CSV ---
    CsvStatus cs;
    cs = csv_open(path_out, &fp_out);
    if (cs != CSV_OK) {
        fprintf(stderr, "Erreur: impossible d'ouvrir le fichier de sortie '%s' (code %d).\n", path_out, cs);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    // --- Lecture des séquences en mémoire ---
    char seq_ref[MAX_SEQ_LEN];
    char seq[MAX_SEQ_LEN];

    size_t n_ref = 0;
    size_t n_seq = 0;

    fs = fasta_read_bases(fp_ref, seq_ref, MAX_SEQ_LEN, &n_ref);
    if (fs != FASTA_OK) {
        fprintf(stderr, "Erreur: échec de la lecture des bases de la séquence de référence (code %d).\n", fs);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    fs = fasta_read_bases(fp_seq, seq, MAX_SEQ_LEN, &n_seq);
    if (fs != FASTA_OK) {
        fprintf(stderr, "Erreur: échec de la lecture des bases de la séquence à comparer (code %d).\n", fs);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    // Vérifier que les longueurs sont identiques
    if (n_ref != n_seq) {
        fprintf(stderr, "Erreur: les séquences ont des longueurs différentes (%zu vs %zu).\n", n_ref, n_seq);
        close_all(fp_ref, fp_seq, fp_out);
        return -1;
    }

    // Vérifier l'alphabet et comparer les séquences
    for (size_t i = 0; i < n_ref; ++i) {
        if (!is_valid_base(seq_ref[i])) {
            fprintf(stderr, "Erreur: caractère invalide '%c' à la position %zu dans la séquence de référence.\n", seq_ref[i], i);
            close_all(fp_ref, fp_seq, fp_out);
            return -1;
        }
        if (!is_valid_base(seq[i])) {
            fprintf(stderr, "Erreur: caractère invalide '%c' à la position %zu dans la séquence à comparer.\n", seq[i], i);
            close_all(fp_ref, fp_seq, fp_out);
            return -1;
        }
        if (seq_ref[i] != seq[i]) {
            cs = csv_write_mutation(fp_out, name_ref, i, seq_ref[i], seq[i]);
            if (cs != CSV_OK) {
                fprintf(stderr, "Erreur: échec de l'écriture dans le fichier CSV (code %d).\n", cs);
                close_all(fp_ref, fp_seq, fp_out);
                return -1;
            }
        }
    }

    // --- Fermeture des fichiers ---
    close_all(fp_ref, fp_seq, fp_out);

    return 0;
}
